### PR TITLE
test(manager/gomod): migrate to fs.readLocalFile

### PR DESCRIPTION
**Pull request by olegkrivtsov**
## Changes:

Migrate to fs.readLocalFile for Gomod (related to #7049)

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/gomod/artifacts.spec.ts [hunk 1]
```diff
@@ -79,13 +74,15 @@ describe('manager/gomod/artifacts', () => {
     ).toBeNull();
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('returns null if unchanged', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:80]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

is the cast required? If not, please remove all those any casts.

File: lib/manager/gomod/artifacts.spec.ts [hunk 2]
```diff
@@ -96,14 +93,15 @@ describe('manager/gomod/artifacts', () => {
     ).toBeNull();
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('returns updated go.sum', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:99]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

You missed some 😕

File: lib/manager/gomod/artifacts.spec.ts [hunk 3]
```diff
@@ -150,15 +149,16 @@ describe('manager/gomod/artifacts', () => {
     ]);
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('supports docker mode without credentials', async () => {
     GlobalConfig.set({ ...adminConfig, binarySource: 'docker' });
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:156]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 4]
```diff
@@ -169,15 +169,16 @@ describe('manager/gomod/artifacts', () => {
     ).not.toBeNull();
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('supports global mode', async () => {
     GlobalConfig.set({ ...adminConfig, binarySource: 'global' });
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:176]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 5]
```diff
@@ -222,13 +224,13 @@ describe('manager/gomod/artifacts', () => {
         matchHost: 'github.enterprise.com',
       },
     ]);
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:228]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 6]
```diff
@@ -263,13 +265,13 @@ describe('manager/gomod/artifacts', () => {
         matchHost: 'gitlab.enterprise.com',
       },
     ]);
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:269]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 7]
```diff
@@ -306,13 +308,13 @@ describe('manager/gomod/artifacts', () => {
         matchHost: 'https://gitlab.enterprise.com/repo2',
       },
     ]);
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:312]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 8]
```diff
@@ -352,13 +354,13 @@ describe('manager/gomod/artifacts', () => {
         matchHost: 'gitlab.enterprise.com',
       },
     ]);
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:358]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 9]
```diff
@@ -402,13 +404,13 @@ describe('manager/gomod/artifacts', () => {
...
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:408]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 10]
```diff
@@ -453,13 +455,13 @@ describe('manager/gomod/artifacts', () => {
         hostType: 'npm',
       },
     ]);
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:459]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 11]
```diff
@@ -486,16 +488,16 @@ describe('manager/gomod/artifacts', () => {
   it('supports docker mode with goModTidy', async () => {
     GlobalConfig.set({ ...adminConfig, binarySource: 'docker' });
     hostRules.find.mockReturnValueOnce({});
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:492]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 12]
```diff
@@ -509,11 +511,12 @@ describe('manager/gomod/artifacts', () => {
...
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('catches errors', async () => {
     const execSnapshots = mockExecAll(exec);
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
-    fs.outputFile.mockImplementationOnce(() => {
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:518]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 13]
```diff
@@ -533,17 +536,18 @@ describe('manager/gomod/artifacts', () => {
     ]);
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('updates import paths with gomodUpdateImportPaths', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:542]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 14]
```diff
@@ -563,16 +567,17 @@ describe('manager/gomod/artifacts', () => {
     ]);
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('skips updating import paths with gomodUpdateImportPaths on v0 to v1', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:573]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 15]
```diff
@@ -591,17 +596,18 @@ describe('manager/gomod/artifacts', () => {
     ]);
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('skips gomodTidy without gomodUpdateImportPaths on major update', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:602]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 16]
```diff
@@ -617,17 +623,18 @@ describe('manager/gomod/artifacts', () => {
     ).toMatchSnapshot();
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('does not execute go mod tidy when none of gomodTidy and gomodUpdateImportPaths are set', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:629]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 17]
```diff
@@ -642,17 +649,18 @@ describe('manager/gomod/artifacts', () => {
     ).toMatchSnapshot();
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('updates import paths with specific tool version from constraint', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:655]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 18]
```diff
@@ -675,17 +683,18 @@ describe('manager/gomod/artifacts', () => {
     ]);
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('updates import paths with latest tool version on invalid version constraint', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:689]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

File: lib/manager/gomod/artifacts.spec.ts [hunk 19]
```diff
@@ -708,16 +717,17 @@ describe('manager/gomod/artifacts', () => {
     ]);
     expect(execSnapshots).toMatchSnapshot();
   });
+
   it('skips updating import paths for gopkg.in dependencies', async () => {
-    fs.readFile.mockResolvedValueOnce('Current go.sum' as any);
-    fs.readFile.mockResolvedValueOnce(null as any); // vendor modules filename
+    fs.readLocalFile.mockResolvedValueOnce('Current go.sum');
+    fs.readLocalFile.mockResolvedValueOnce(null as any); // vendor modules filename
```
Comment (viceice) [inline lib/manager/gomod/artifacts.spec.ts:723]:
```suggestion
    fs.readLocalFile.mockResolvedValueOnce(null); // vendor modules filename
```

**Review by viceice — COMMENTED** [review]
😕

File: lib/manager/gomod/artifacts.spec.ts (hunk 2 above)
Comment (olegkrivtsov) [reply lib/manager/gomod/artifacts.spec.ts:99]: Hi @viceice sorry, I didn't have a chance to fix this yet, since I struggle with `gradle-wrapper` right now.

File: lib/manager/gomod/__snapshots__/artifacts.spec.ts.snap [hunk 20]
```diff
@@ -31,7 +31,6 @@ Array [
         "GOFLAGS": "-modcacherw",
         "GONOPROXY": "noproxy.example.com/*",
         "GONOSUMDB": "1",
-        "GOPATH": "/tmp/renovate/cache/others/go",
```
Comment (viceice) [inline lib/manager/gomod/__snapshots__/artifacts.spec.ts.snap:34]: Any idea why this changes?

Comment (olegkrivtsov) [reply lib/manager/gomod/__snapshots__/artifacts.spec.ts.snap:34]: This changed because previously we didn't mock `ensureCacheDir()`, but now we mock it. So it returns `undefined` and GOPATH is undefined.

**Review by viceice — APPROVED** [review]
